internal/charmstore/archive.go

```diff
@@ -4,16 +4,229 @@
 package charmstore // import "gopkg.in/juju/charmstore.v5-unstable/internal/charmstore"
 
 import (
+	"archive/zip"
 	"bytes"
 	"io"
 	"os"
+	"path"
+	"strings"
 
+	"github.com/juju/utils"
 	"gopkg.in/errgo.v1"
 	"gopkg.in/juju/charm.v6-unstable"
+	"gopkg.in/juju/charmrepo.v2-unstable/csclient/params"
+	"gopkg.in/mgo.v2/bson"
 
 	"gopkg.in/juju/charmstore.v5-unstable/internal/blobstore"
+	"gopkg.in/juju/charmstore.v5-unstable/internal/mongodoc"
+	"gopkg.in/juju/charmstore.v5-unstable/internal/router"
 )
 
+// Blob represents a blob of data from the charm store.
+type Blob struct {
+	blobstore.ReadSeekCloser
+
+	// Size holds the total size of the blob.
+	Size int64
+
+	// Hash holds the hash checksum of the blob.
+	Hash string
+}
+
+var preV5ArchiveFields = []string{
+	"size",
+	"blobhash",
+	"blobname",
+	"prev5blobhash",
+	"prev5blobsize",
+}
+
+// OpenBlob returns the blob associated with the given URL.
+func (s *Store) OpenBlob(id *router.ResolvedURL) (*Blob, error) {
+	return s.openBlob(id, false)
+}
+
+// OpenBlob returns the blob associated with the given URL.
+// As required by pre-v5 versions of the API, it will return a blob
+// with a hacked-up metadata.yaml that elides the Series field.
+func (s *Store) OpenBlobPreV5(id *router.ResolvedURL) (*Blob, error) {
+	return s.openBlob(id, true)
+}
+
+func (s *Store) openBlob(id *router.ResolvedURL, preV5 bool) (*Blob, error) {
+	entity, err := s.FindEntity(id, FieldSelector(preV5ArchiveFields...))
+	if err != nil {
+		return nil, errgo.Mask(err, errgo.Is(params.ErrNotFound))
+	}
+	r, size, err := s.BlobStore.Open(entity.BlobName)
+	if err != nil {
+		return nil, errgo.Notef(err, "cannot open archive data for %s", id)
+	}
+	hash := entity.BlobHash
+
+	if entity.PreV5BlobHash != entity.BlobHash && preV5 {
+		// The v5 blob is different so we open the blob suffix that
+		// contains the metadata hack.
+		r2, size2, err := s.BlobStore.Open(preV5CompatibilityBlobName(entity.BlobName))
+		if err != nil {
+			r.Close()
+			return nil, errgo.Notef(err, "cannot find pre-v5 hack blob")
+		}
+		r = newMultiReadSeekCloser(r, r2)
+		size += size2
+		hash = entity.PreV5BlobHash
+	}
+	return &Blob{
+		ReadSeekCloser: r,
+		Size:           size,
+		Hash:           hash,
+	}, nil
+}
+
+type multiReadSeekCloser struct {
+	readers []blobstore.ReadSeekCloser
+	io.ReadSeeker
+}
+
+func newMultiReadSeekCloser(readers ...blobstore.ReadSeekCloser) blobstore.ReadSeekCloser {
```

Comment:
I *think* it should be obvious enough from the signature and the similarity to io.MultiReader.